### ISSUE TITLE
Huge battery consumption during Android Sleep if Azahar was active when entering Sleep

### ISSUE THREAD

**Issue by Anutim** [Lime3DS/Lime-3DS-Emulator#519]
### Is there an existing issue for this?

- [X] I have searched the existing issues

### Affected Build(s)

2118.2

### Description of Issue

If you enter Androids Sleep mode, there's a huge battery drain compared to other apps. (Lime3DS consumes like 4+% per hour, whilst most other apps I've tried are <1% per hour.)

This behaviour is exhibited on all emulators I've tried that are based on the same foundation as Lime3DS is based on, incase that helps narrow down what could be the issue.

Not sure if this is related/helps, but if you press Pause emulation, then Sleep.
Once you 'unsleep', the games already playing even though the 'Pause emulation' button still has the text 'Resume emulation'.

### Expected Behavior

<1% battery consumption per hour, like pretty much all emulators I've tried apart from like Citra, Yuzu etc.

### Reproduction Steps

I have an Retroid Pocket 4 Pro that I've tested this on, but I've talked to people on other devices (Retroid Pocket Mini etc) who also have the issue, so it doesn't seem to be platform specific.

### Log File

No log file as this isn't related to any crash etc.

### System Configuration

CPU: MTK D1100
GPU/Driver: Mali-G77 MP9
RAM: 8GB LPDDR4x (not sure what brand/product line Retroid Pocket 4 Pro uses)
OS: Android 13

**Comment by EthanDonley** [comment]
I happen to come across this issue as well on the AYN Odin 2/Retroid Pocket 4 Pro/Retroid Pocket Mini/Retroid Pocket 2S. So it's a thing on Android 10/11/13. There was a pretty old build of Citra MMJ that I believe didn't have this problem but later updates of MMJ eventually brought this issue back in.

**Comment by Anutim** [comment]
> I happen to come across this issue as well on the AYN Odin 2/Retroid Pocket 4 Pro/Retroid Pocket Mini/Retroid Pocket 2S. So it's a thing on Android 10/11/13. There was a pretty old build of Citra MMJ that I believe didn't have this problem but later updates of MMJ eventually brought this issue back in.

Do you know sort of when this old build of MMJ was made? I don't mind trying several different versions to test.

**Comment by kalebepalacio** [comment]
Just remove lime3ds from multitasking (recent). Then I finish using it and do this. It works.

**Comment by Anutim** [comment]
> Just remove lime3ds from multitasking (recent). Then I finish using it and do this. It works.

Not entirely sure what you mean by this, but if it's about preventing it from running in the background, it's not related to the post.

The premise is: if you have a game running, then put your device to sleep, then wake it up to resume playing your game. It will have consumed a lot more battery than it should have.

**Comment by kalebepalacio** [comment]
I apologize. I now understand that you are talking about portable devices with Android intended for emulation.

**Comment by ghost** [comment]
> The premise is: if you have a game running, then put your device to sleep, then wake it up to resume playing your game. It will have consumed a lot more battery than it should have.

That's not as easy to activate as you might think on Android

**Comment by Anutim** [comment]
> That's not as easy to activate as you might think on Android

Sure, I mean I have no idea what goes into it. But I'd for sure consider it the biggest problem with using the emulator on Android as is.

Savestates help the situation a lot though, but it's still pretty clunky to savestate->close emu/open emu->load savestate.
Especially if you're like playing while waiting for someone. I've run into several instances playing MM3D where I'm waiting for someone and play a bit, then when they show up I just instantly hit the sleep button and forget about it and gg battery dead lol.

> Not sure if this is related/helps, but if you press Pause emulation, then Sleep.
> Once you 'unsleep', the games already playing even though the 'Pause emulation' button still has the text 'Resume emulation'.

What about this btw? It's not related is it?

**Comment by ghost** [comment]
> What about this btw? It's not related is it?

Oh that completely screws with the pause emulator feature i can reproduce it
Maybe fixing this could fix your issues as well although i still am not sure how well it will handle deep sleep on Android

**Comment by Anutim** [comment]
> Oh that completely screws with the pause emulator feature i can reproduce it Maybe fixing this could fix your issues as well although i still am not sure how well it will handle deep sleep on Android

Well, if you do tinker with it and need a tester, let me know.

**Comment by EthanDonley** [comment]
Just want to point out that with Azahar's release that this is still an issue!

**Comment by rcerrejon** [comment]
The same thing has happened to me in Odin 2, when the device is in sleep, it consumes almost as much as if it was playing.
Maybe it's because for example “PARTIAL_WAKE_LOCK” keeps the cpu active or a “foreground service” running constantly? Just an idea.

**Comment by Anutim** [comment]
Just did a quick test to see if it's still an issue since the bot seems to want to get rid of it lol.
And yes, tried on my RP5 and in 2 hours 4% gone, which while it isn't as bad as my original tests on my RP4, it's still far from normal for an android app.

Though, should be mentioned the battery curve is pretty goofy on the firmware version I'm on and I'm ~50% which is around where it tends to move slowest, so would assume it's possibly worse than the 2 hours 4% mention.

**Comment by NicLinley** [comment]
I think I’m experiencing this problem on my AYN Thor. When azahar is open and in the background or the device should be in sleep mode the amount of battery lost is noticeably a lot more than with other apps and emulators and also the back side is warm to the touch while in sleep mode.

**Comment by Anutim** [comment]
> I think I’m experiencing this problem on my AYN Thor.

Yeah, can confirm. Just tried it on the Thor, and in 2 hours, I lost 7%, which lines up very similarly to the initial test I posted of my RP4.

**Comment by DavidRGriswold** [comment]
I can try to look into this but it may be a little bit of a Rock and a Hard Place thing where making it behave better in sleep might also make it more likely the game will just die while sleeping. But I can do some experimenting to see what I can uncover.

**Comment by Anutim** [comment]
> I can try to look into this but it may be a little bit of a Rock and a Hard Place thing where making it behave better in sleep might also make it more likely the game will just die while sleeping. But I can do some experimenting to see what I can uncover.

If you do figure out an improvement, a PR for MMJ would be great too, as I'd imagine the issue there would be the same thing as with Azahar.

**Comment by PabloMK7** [comment]
> > I can try to look into this but it may be a little bit of a Rock and a Hard Place thing where making it behave better in sleep might also make it more likely the game will just die while sleeping. But I can do some experimenting to see what I can uncover.
> 
> If you do figure out an improvement, a PR for MMJ would be great too, as I'd imagine the issue there would be the same thing as with Azahar.

We cannot PR anything to MMJ because it's license state is unclear.

**Comment by manzaloros** [comment]
I also experience the high battery drain in Android sleep mode.

**Comment by ClementNerma** [comment]
Still a problem to this day with version 2124.3.

**Comment by TMBL-DEV** [comment]
Still have this issue, battery drains when putting device to sleep.
